broadinstitute/single_cell_portal_core
# Adding lightweight A/B test framework (SCP-4872)

#### BACKGROUND
Running A/B tests is an established practice for measuring the impact of a UI/UX update in a controlled manner.  It allows side-by-side comparisons of a control group with an updated UX without needing to establish or collect baseline metrics ahead of time.

#### CHANGES
This update introduces a lightweight A/B test framework based off of the `FeatureFlag` model.  As these are already used to control UI/UX updates, we have an established practice of changing the UI based off of flag settings.  This adds a `enable_ab_test` boolean flag that will govern whether or not an A/B test is being run on that `FeatureFlag`.  In addition, the `AbTestSession` model will track group assignments for all enabled flags for both unauthenticated and authenticated users via the `user_id` UUID cookie that is already present in the SCP DOM.

Once an A/B test is enabled, users will be randomly divided into the `control` and `experiment` groups as they visit SCP.  These assignments are sticky - they persist whether the user is signed in or not (provided they don't clear their cookies), and will remain the same for the entire duration of the A/B test on that feature.  These assignments will be reported globally to Mixpanel via the `window.SCP` object under a new field called `aBFeatureTests`.  For example, if an A/B test was enabled for the feature flag called `progressive_loading` (which governs whether or not to load static images of gene expression scatter plots prior to making API requests), users would be divided into the groups `progressive-loading-group-control` and `progressive-loading-group-experiment`.

Also - and importantly - since `FeatureFlag` values can be overridden at the user level, if any given user has an override in place for a feature that is being A/B tested, that particular group assignment will not be reported.  This is because the state of the override may result in that user being show the opposite UX from what the group assignment would see.

NOTE: This update does not enable any A/B tests by default, or handle updating the UI for those tests.  This only handles group assignment/reporting.

#### MANUAL TESTING
1. Boot as normal and open the home page
2. Open the DevTools -> Console and inspect the `window.SCP` object to see that `aBFeatureTests` is empty:
![singed_out_no_session](https://user-images.githubusercontent.com/729968/217061385-439141b9-e365-4871-88b4-e5112e5d0528.png)
3. In a separate Rails console session, get a list of `FeatureFlag` names where you have not configured an override for your user account (in this case, there were 3 found - yours may be different):
```
email = '<your email>'
user = User.find_by(email:)
flag_names = FeatureFlag.pluck(:name) - user.configured_feature_flags.keys
=> ["clientside_validation", "reference_image_upload", "progressive_loading"]   
```
4. Enable A/B tests for these feature flags:
```
flag_names.each do |name|
  feature_flag = FeatureFlag.find_by(name:)
  feature_flag.update(enable_ab_test: true)
end
```
5. Reload the page, and note that there are now group assignments present (the names/values will be different depending on which flags were enabled, but you should see one for every flag enabled):
![signed_out_test_sessions](https://user-images.githubusercontent.com/729968/217060221-99877346-50ab-4d7e-b67b-44df746987d8.png)
6. Sign in and confirm that your group assignments do not change
7. Set a flag override for your account for each flag, then reload the page and confirm no group assignments are being reported:
```
flag_names.each do |name|
  user.set_flag_option(name, true)
end
```
![singed_in_no_sessions](https://user-images.githubusercontent.com/729968/217060483-822c8497-bc68-4283-a903-ceaeb9fbe529.png)
8. Unset the flag overrides, reload the page and confirm that you are back to the same group assignments as before:
```
flag_names.each do |name|
  user.remove_flag_option(name)
end
```
![signed_out_test_sessions](https://user-images.githubusercontent.com/729968/217061135-c9c7fc57-07f9-4ac4-aa8a-d14a762a3cbb.png)
9. (Optional) Clean up to remove all test assignments:
```
FeatureFlag.where(enable_ab_test: true).update_all(enable_ab_test: false)
AbTestSession.destroy_all
```
